File: src/main/java/net/csibio/aird/compressor/bytecomp/ZstdWrapper.java
package net.csibio.aird.compressor.bytecomp;

import com.github.luben.zstd.Zstd;
import com.github.luben.zstd.ZstdDictCompress;
import com.github.luben.zstd.ZstdDictDecompress;
import net.csibio.aird.enums.ByteCompType;

import java.util.Arrays;

public class ZstdWrapper implements ByteComp {

    @Override
    public String getName() {
        return ByteCompType.Zstd.getName();
    }

    @Override
    public byte[] encode(byte[] input) {
        return Zstd.compress(input);
    }

    @Override
    public byte[] decode(byte[] input) {
        int size = (int) Zstd.decompressedSize(input);
        byte[] array = new byte[size];
        try {
            Zstd.decompress(array, input);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return array;
    }

    @Override
    public byte[] decode(byte[] input, int offset, int length) {
        try {
            return decode(Arrays.copyOfRange(input, offset, offset + length));
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public byte[] encode(byte[] input, byte[] dict) {
        return Zstd.compress(input, new ZstdDictCompress(dict, 3));
    }

    public byte[] decode(byte[] input, byte[] dict) {
        int size = (int) Zstd.decompressedSize(input);
        return Zstd.decompress(input, new ZstdDictDecompress(dict), size);
    }

    public byte[] decode(byte[] input, int offset, int length, byte[] dict) {
        return decode(Arrays.copyOfRange(input, offset, offset + length), dict);
    }
}

File: CSharpSDK/Compressor/ByteComp/ZstdWrapper.cs
/*
 * Copyright (c) 2020 CSi Studio
 * AirdSDK and AirdPro are licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2. 
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2 
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.  
 * See the Mulan PSL v2 for more details.
 */

using AirdSDK.Enums;
using System;
using ZstdNet;

namespace AirdSDK.Compressor
{
    public class ZstdWrapper : ByteComp
    {
        //使用zstd将byte数组压缩
        public override string getName()
        {
            return ByteCompType.Zstd.ToString();
        }

        public override byte[] encode(byte[] data)
        {
            var compressor = new ZstdNet.Compressor();
            var compressedData = compressor.Wrap(data);
            return compressedData;
        }

        public override byte[] decode(byte[] data)
        {
            var decompressor = new Decompressor();
            var decompressedData = decompressor.Unwrap(data);
            return decompressedData;
        }

        public override byte[] decode(byte[] input, int offset, int length)
        {
            byte[] result = new byte[length];
            Array.Copy(input, offset, result, 0, length);

            var decompressor = new Decompressor();
            var uncompressed = decompressor.Unwrap(result);

            return uncompressed;
        }
    }
}